Resource Based Permission - Angular UI Issue 24299

fahrigedik opened the pull request:
### Description

Resolves #24299  (write the related issue number if available)

### How to test it?

you have to change test-branch-24299
you have to run dbmigrator and hostwithids on app templates
you have to run dev-app

fahrigedik committed:
Add resource permission management component

Introduces ResourcePermissionManagementComponent with UI and logic for managing resource-based permissions. Updates models and service to support resource permission APIs, and exports the new component in the index.

fahrigedik committed:
Refactor resource permission management to use extensible table

Moved resource-permission-management component files into a dedicated folder and updated imports. Replaced the manual table implementation with the ExtensibleTableComponent for resource permissions, added client-side pagination, and introduced extension points for entity property configuration. Added new defaults, services, and tokens to support extensibility and improved maintainability.

fahrigedik committed:
Add new localization keys for permission management

Added new translation keys for 'UpdatePermission', 'NoPermissionsAssigned', 'SelectProvider', 'SearchProviderKey', and 'Provider' across all supported languages in the permission management module. This enhances localization support for new UI features related to provider selection and permission assignment.

fahrigedik committed:
Add new localization keys for permission errors

Added 'ErrorLoadingPermissions' and 'PleaseSelectProviderAndPermissions' keys to all language JSON files in the permission management module to support new error and prompt messages in the UI.

fahrigedik committed:
Improve provider key search UX and add delete confirmation

Enhances the provider key search input with focus/blur handling and a dropdown for results, improving user experience. Adds a confirmation dialog before deleting a permission, preventing accidental deletions. Refactors related logic for clarity and better state management.

sumeyyeKurtulus commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.ts:
```diff
@@ -0,0 +1,389 @@
+    protected readonly confirmationService = inject(ConfirmationService);
+    readonly list = inject(ListService);
+
+    @Input() resourceName!: string;
```
It is better to use the latest `input` and `output` signal functions instead.

sumeyyeKurtulus commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.ts:
```diff
@@ -0,0 +1,389 @@
+        }
+    }
+
+    toggleAllPermissions(selectAll: boolean) {
```
I think, we can improve the readability here like this
```ts
toggleAllPermissions(selectAll: boolean) {
        let toggleAction = selectAll ? this.permissionsWithProvider().map(p => p.name || '') : [];
        if (this.viewMode() === 'add') {
            toggleAction = this.permissionDefinitions().map(p => p.name || '');
        }
        this.selectedPermissions.set(toggleAction);
    }
```

sumeyyeKurtulus commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.html:
```diff
@@ -0,0 +1,165 @@
+                }
+            </div>
+
+            <div class="position-relative mb-3">
```
We can extract this and create generic component to behave like a lookup search component. You can check this as a reference https://github.com/volosoft/volo/tree/dev/abp/npm/ng-packs/packages/commercial-ui/src/lib/components/lookup

sumeyyeKurtulus commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.html:
```diff
@@ -0,0 +1,165 @@
+            </button>
+        </div>
+
+        <ng-template #actionsTemplate let-row>
```
These buttons does not seem configured to be a dropdown as an action button. We should try the bootstrap dropdown if possible.

sumeyyeKurtulus committed:
update: sv.json file for an extra key

sumeyyeKurtulus committed:
update: proxy generation file for the permission management

fahrigedik committed:
Refactor resource permission management to use state service

Introduces ResourcePermissionStateService to centralize and manage state for resource permission management. Refactors resource-permission-management.component and related templates to use the new state service, and extracts form, list, and search UI into dedicated components for improved modularity and maintainability.

fahrigedik committed:
Refactor components to use external HTML templates

Moved inline templates to separate HTML files for permission-checkbox-list, provider-key-search, resource-permission-form, and resource-permission-list components. Updated component decorators to reference the new template files, improving maintainability and separation of concerns.

erdemcaygor commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.ts:
```diff
@@ -0,0 +1,389 @@
+        this.searchSubject.pipe(
+            debounceTime(300),
+            distinctUntilChanged()
+        ).subscribe(filter => {
```
The subscription should be unsubscribed when the page is destroyed.

fahrigedik committed:
Add reusable lookup-search component and refactor provider key search

Introduces a generic LookupSearchComponent for reusable lookup/search UI in Angular. Refactors the provider key search in permission management to use the new component, simplifying its logic and improving maintainability. Updates tsconfig paths to include the new lookup component package.

erdemcaygor commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.html:
```diff
@@ -0,0 +1,165 @@
+<abp-modal [(visible)]="visible" [busy]="modalBusy()" [options]="{ size: 'xl', scrollable: false }">
+    <ng-template #abpHeader>
+        <h5 class="modal-title">
+            @if (viewMode() === 'edit') {
```
Maybe we can define enum for ViewModes instead of using hardcoded string. Example;

```
export enum ViewModeValues {
  List = 'list',
  Add = 'add',
  Edit = 'edit'
}

```

fahrigedik committed:
Refactor view modes to use enum in permission management

Replaced string literals for view modes with the new eResourcePermissionViewModes enum across components and services. This improves type safety and maintainability. Added the enum definition in a new file.

sumeyyeKurtulus commented on npm/ng-packs/packages/components/lookup/src/lib/lookup-search.component.ts:
```diff
@@ -0,0 +1,139 @@
+        this.searchSubject.pipe(
+            debounceTime(this.debounceTime()),
+            distinctUntilChanged(),
+            takeUntil(this.destroy$)
```
We can use `takeUntilDestroyed` instead. By this means, we will not need this un-subscription on destroy 

```tsx
  ngOnDestroy() {
    this.destroy$.next();
    this.destroy$.complete();
  }
```

The same applies for this component :: `npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/provider-key-search/provider-key-search.component.ts`

Here is a quick reference https://www.reddit.com/r/angular/comments/12vbfbd/angular_16_takeuntildestroyed_operator/

sumeyyeKurtulus commented on npm/ng-packs/packages/components/lookup/src/lib/lookup-search.component.ts:
```diff
@@ -0,0 +1,139 @@
+        ).subscribe({
+            next: results => {
+                this.searchResults.set(results);
+                this.isLoading.set(false);
```
We can set the load false on finalize inside pipe to remove the duplication.

sumeyyeKurtulus commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.ts:
```diff
@@ -0,0 +1,206 @@
+    private previousVisible = false;
+
+    constructor() {
+        effect(() => {
```
We can make this effect work more efficiently. Here is a quick example 

```tsx
    effect(() => {
      // Read all input values to track them
      const resourceName = this.resourceName();
      const resourceKey = this.resourceKey();
      const resourceDisplayName = this.resourceDisplayName();

      // Use untracked to prevent this effect from tracking writes to state signals
      // This prevents circular dependencies and reduces unnecessary re-executions
      untracked(() => {
        // Only update if values have actually changed (signals optimize this internally)
        this.state.resourceName.set(resourceName);
        this.state.resourceKey.set(resourceKey);
        this.state.resourceDisplayName.set(resourceDisplayName);
      });
    });
 ```

sumeyyeKurtulus committed:
update: self-closing tag usage

sumeyyeKurtulus committed:
add: index file for enums

sumeyyeKurtulus committed:
update: auto-save changes

erdemcaygor commented on npm/ng-packs/packages/components/lookup/src/lib/lookup-search.component.html:
```diff
@@ -0,0 +1,46 @@
+
+    @if (showDropdown() && !disabled()) {
+    <div class="abp-lookup-dropdown list-group position-absolute w-100 shadow"
+        style="z-index: 1050; max-height: 200px; overflow-y: auto;">
```
it's better to use class instead of inline styles

erdemcaygor commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/permission-checkbox-list/permission-checkbox-list.component.html:
```diff
@@ -0,0 +1,25 @@
+    <div class="border rounded p-3" style="max-height: 300px; overflow-y: auto;">
+        @for (perm of permissions(); track perm.name) {
+        <div class="form-check">
+            <input class="form-check-input" type="checkbox" [id]="'perm-' + idPrefix() + '-' + perm.name"
```
we can use string interpolation. it's more readable.

`id="perm-{{ idPrefix() }}-{{ perm.name }}"`

erdemcaygor commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/permission-checkbox-list/permission-checkbox-list.component.html:
```diff
@@ -0,0 +1,25 @@
+            {{ 'AbpPermissionManagement::GrantAllResourcePermissions' | abpLocalization }}
+        </label>
+    </div>
+    <div class="border rounded p-3" style="max-height: 300px; overflow-y: auto;">
```
it's better to use class instead of inline styles

erdemcaygor commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.ts:
```diff
@@ -0,0 +1,206 @@
+
+    readonly visible = model<boolean>(false);
+
+    private previousVisible = false;
```
we can use signal for `previousVisible`

erdemcaygor commented on npm/ng-packs/packages/permission-management/src/lib/components/resource-permission-management/resource-permission-management.component.ts:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    ngOnInit() {
+        this.list.maxResultCount = 10;
```
maybe we can define constant variable for maxResultCount instead of using hardcoded value

fahrigedik committed:
Refactor lookup and permission components, add styles

Refactored lookup-search and permission-checkbox-list components to use SCSS for dropdown and list container styling. Improved resource permission management component with signals and untracked for state updates, and replaced manual destroy logic with Angular's DestroyRef and takeUntilDestroyed. Updated HTML templates for better readability and accessibility.
